### PR TITLE
Use barcodex to extract UMIs

### PR DESCRIPTION
**Pull request by apmasell**
This adds UMI extraction support to the BCL2FASTQ workflow. I've added a lot of people to this PR, but I'm asking for specific things from each of you:

- C++ code readability: Angie and Dillan
- barcodex logic/output verification: Richard and Larry
- WDL changes: Angie and Mike

### CODE REVIEW

**Comment by apmasell** [comment]
retest this please

File: wrapper/main.cpp [hunk 1]
```diff
@@ -167,6 +178,249 @@ class Sample {
...
+// the sample sheet if multiple barcodes) and needs UMI extraction done on the
+// output
+class UmiSample : public Sample {
+public:
+  UmiSample(Json::ArrayIndex id_, const std::string &name_, int barcodes_,
+            bool is_inline_, const std::string &acceptableUmiList_,
+            const Json::Value &patterns_)
+      : id(id_), acceptableUmiList(acceptableUmiList_), barcodes(barcodes_),
+        is_inline(is_inline_), name(name_) {
+    if (patterns_.isObject()) {
```
Comment (asofmosq) [inline wrapper/main.cpp:248]: What's happening here?

File: wrapper/main.cpp [hunk 2]
```diff
@@ -167,6 +178,249 @@ class Sample {
...
+          pair["right"] = std::move(attributes);
+          record["fastqs"] = std::move(pair);
+          record["name"] = name;
+          output.append(std::move(record));
+          // We're going to need to update these later, so hold references to
+          // them that we can pass to the other thread.
+          readCountOutput.push_back((*--output.end())["fastqs"]["right"]);
+        }
+
+        Json::Value record(Json::objectValue);
```
Comment (asofmosq) [inline wrapper/main.cpp:364]: Perhaps include a comment here that this is for the base file?

File: wrapper/main.cpp [hunk 3]
```diff
@@ -167,6 +178,249 @@ class Sample {
...
+        // tell barcodex about them
+        std::stringstream samplesheetpattern;
+        samplesheetpattern << "sample" << id << "_" << barcode << "_S*_R"
+                           << read << "_001.fastq.gz";
+        std::stringstream parameter;
+        parameter << "--r" << read << "_in";
+
+        for (auto &fastq : fastqs) {
+          switch (fnmatch(samplesheetpattern.str().c_str(), fastq.c_str(),
+                          FNM_PATHNAME)) {
```
Comment (asofmosq) [inline wrapper/main.cpp:392]: what's FNM?

**Comment by apmasell** [comment]
retest this please

**Comment by apmasell** [comment]
All tests pass and data has been verified.

**Pull request merged**